# Bài tập 1: Sinh viên cho ví dụ về phép cộng, trừ hai ma trận trong numpy

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [1]:
import numpy as np
a = np.array([-8, 15])
b = np.array([9, 2])
# Cộng
y = np.add(a, b)
y

array([ 1, 17])

In [2]:
# Trừ
z = np.subtract(a, b)
z

array([-17,  13])

# Bài tập 2: Pandas

## Đọc CSV thành Dataframe và hiển thị

In [ ]:
import pandas as pd
df = pd.read_csv("/content/SAMPLE-CSV.csv", delimiter=';')
df

X   Y   Z
0  78  84  86
1  85  94  97
2  80  83  73
3  96  94  96
4  86  86  83

## Hãy chuyển index mặc định thành giá trị cột id

In [ ]:
df['id'] = df.index
df.set_index('id', inplace=True)
df

X   Y   Z
id            
0   78  84  86
1   85  94  97
2   80  83  73
3   96  94  96
4   86  86  83

## Sắp xếp dữ liệu theo nhiều cột (sort)

In [ ]:
df.sort_index(axis=0, ascending=False)

X   Y   Z
id            
4   86  86  83
3   96  94  96
2   80  83  73
1   85  94  97
0   78  84  86

## Chọn một cột cụ thể và hiển thị chúng

In [ ]:
df['X']

id
0    78
1    85
2    80
3    96
4    86
Name: X, dtype: int64

## Chọn 2 hàng đầu tiên và hiển thị chúng

In [ ]:
df.head(2)

X   Y   Z
id            
0   78  84  86
1   85  94  97

## Hãy chọn một hàng dựa trên một điều kiện giá trị của cột

In [ ]:
df.loc[df['X'] > 80]

X   Y   Z
id            
1   85  94  97
3   96  94  96
4   86  86  83

## Thay đổi một vài giá trị thành NaN ở CSV, sau đó đọc lên thành Dataframe và thay thế chúng bằng giá trị 0

In [ ]:
# replace some values in df to NaN
import numpy as np
df.loc[df['X'] < 85, 'X'] = np.nan
df

X   Y   Z
id              
0    NaN  84  86
1   85.0  94  97
2    NaN  83  73
3   96.0  94  96
4   86.0  86  83

## Ở cột Z chuyển giá trị lớn hơn 90 là True và nhỏ hơn là False trong Dataframe

In [ ]:
# convert values if Z column to True if greater than 90 else Fasle
df['Z'] = df['Z'].apply(lambda x: True if x > 90 else False)
df

X   Y      Z
id                 
0    NaN  84  False
1   85.0  94   True
2    NaN  83  False
3   96.0  94   True
4   86.0  86  False

## Chuyển Dataframe trên thành 2 Dataframe d1 và d2; d1 chứa cột X và Y, d2 chứa cột Z, cuối cùng d3 là thành quả của nối 2 Dataframe d1 và d2

In [ ]:
d1 = df[['X', 'Y']]
d2 = df[['Z']]
d3 = pd.concat([d1, d2], axis=1)
d3

X   Y      Z
id                 
0    NaN  84  False
1   85.0  94   True
2    NaN  83  False
3   96.0  94   True
4   86.0  86  False

## Dùng tính năng thống kê hãy hiển thị kết quả thống kê các giá trị thuộc tính của Dataframe

In [ ]:
df.describe()

X          Y
count   3.000000   5.000000
mean   89.000000  88.200000
std     6.082763   5.403702
min    85.000000  83.000000
25%    85.500000  84.000000
50%    86.000000  86.000000
75%    91.000000  94.000000
max    96.000000  94.000000

# Bài tập về nhà 3: Sử dụng Linear Regression

### Scikit-learning

In [ ]:
import numpy as np
from sklearn.linear_model import LinearRegression
# X is a matrix that represents the training dataset
# y is a vector of weights, to be associated with input dataset
X = np.array([[3], [5], [7], [9], [11]]).reshape(-1, 1)
y = [8.0, 9.1, 10.3, 11.4, 12.6]
lreg_model = LinearRegression()
lreg_model.fit(X, y)
# New data (unseen before)
new_data = np.array([[13]])
print('Model Prediction for new data: $%.2f' % lreg_model.predict(new_data)[0])

Model Prediction for new data: $13.73


## Tìm hiểu thực hiện lại ví dụ dùng mô hình Linear Regression trong thư viện scikit-learning bằng các thư viện sau:

### TensorFlow

In [ ]:
import numpy as np
import tensorflow as tf

X = np.array([[3], [5], [7], [9], [11]], dtype=np.float32)
y = np.array([8.0, 9.1, 10.3, 11.4, 12.6], dtype=np.float32)
# Weight and bias
W = tf.Variable(0.0, dtype=tf.float32)
b = tf.Variable(0.0, dtype=tf.float32)

optimizer = tf.keras.optimizers.SGD(learning_rate=0.01)

for epoch in range(1000):
    with tf.GradientTape() as tape:
        y_pred = W * X + b
        loss = tf.reduce_mean(tf.square(y - y_pred))
    gradients = tape.gradient(loss, [W, b])
    optimizer.apply_gradients(zip(gradients, [W, b]))

# New data (unseen before)
new_data = np.array([[13]], dtype=np.float32)
prediction = (W * new_data + b).numpy()[0]

print(f"TensorFlow Model Prediction: ${prediction.item():.2f}")

TensorFlow Model Prediction: $10.66


### Keras

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

# Data
X = np.array([[3], [5], [7], [9], [11]], dtype=np.float32)
y = np.array([8.0, 9.1, 10.3, 11.4, 12.6], dtype=np.float32)

model = keras.Sequential([
    keras.layers.Dense(1, input_shape=(1,), activation=None)  # Hồi quy tuyến tính
])
model.compile(optimizer='sgd', loss='mse')

# Train
model.fit(X, y, epochs=1000, verbose=0)

# Test
new_data = np.array([[13]], dtype=np.float32)
prediction = model.predict(new_data)[0][0]
print(f"Keras Model Prediction: ${prediction:.2f}")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step
Keras Model Prediction: $13.96


### PyTorch

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim

# Data
X = torch.tensor([[3], [5], [7], [9], [11]], dtype=torch.float32)
y = torch.tensor([[8.0], [9.1], [10.3], [11.4], [12.6]], dtype=torch.float32)

model = nn.Linear(1, 1)
criterion = nn.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Train
for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = model(X)
    loss = criterion(y_pred, y)
    loss.backward()
    optimizer.step()

# Test
new_data = torch.tensor([[13]], dtype=torch.float32)
prediction = model(new_data).item()
print(f"PyTorch Model Prediction: ${prediction:.2f}")

PyTorch Model Prediction: $13.97


## Cảm	nghĩ	về	việc	dùng	4	thư	viện	này

1.   Scikit-Learn:
Dễ sử dụng nhất vì có sẵn LinearRegression(), chỉ cần gọi .fit() là xong.
Phù hợp cho các bài toán machine learning đơn giản, không cần tối ưu hóa sâu.
Tuy nhiên, không linh hoạt khi muốn tùy chỉnh quá trình huấn luyện hoặc kiểm soát chi tiết thuật toán tối ưu.
2.   TensorFlow:
Mạnh mẽ cho các mô hình deep learning, nhưng hơi cồng kềnh với bài toán hồi quy tuyến tính đơn giản.
Cần tự định nghĩa trọng số, bias, và vòng lặp huấn luyện.
Kiểm soát được quá trình học, nhưng phải hiểu về GradientTape() và apply_gradients(), khiến việc triển khai phức tạp hơn.
3.   Keras:
Giao diện thân thiện hơn vì có thể dùng API cấp cao như Sequential và Dense.
Dễ dùng hơn TensorFlow, nhưng vẫn mạnh mẽ nếu kết hợp với các lớp trừu tượng của nó.
Không tối ưu bằng Scikit-Learn cho bài toán nhỏ, nhưng dễ mở rộng sang mô hình lớn hơn.
4.   PyTorch:
Rất phổ biến trong cộng đồng nghiên cứu, dễ dàng kiểm soát trọng số và quá trình tính gradient.
API trực quan hơn TensorFlow, nhất là với torch.nn và torch.optim.
Nhưng đối với bài toán đơn giản như hồi quy tuyến tính, cách triển khai cũng phức tạp hơn Scikit-Learn.

Tổng kết:
Nếu chỉ cần hồi quy tuyến tính nhanh chóng, Scikit-Learn là lựa chọn tốt nhất. Nếu muốn kiểm soát chi tiết hơn quá trình học, có thể dùng PyTorch hoặc TensorFlow. Còn Keras là lựa chọn trung gian, dễ dùng hơn TensorFlow nhưng vẫn đủ linh hoạt cho các mô hình mở rộng.

# Bài tập 4: Code phát hiện Spam bằng SVMs và Linear Regression

## Bằng SVMs

In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd
import numpy as np
# Đọc file dataset, biến thành các giá trị số
df = pd.read_csv('/content/sms_spam_svm.csv')
y = df.iloc[:, 0].values
y = np.where(y == 'spam', -1, 1)
X = df.iloc[:, [1, 2]].values
# chia dữ liệu thành tập train và tập test
# test_size là 30%
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size=0.3, random_state=0)
svm = SVC(kernel='linear', C=1.0, random_state=0)
svm.fit(X_train, y_train)
y_pred = svm.predict(X_test)
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.3f' % accuracy_score(y_test, y_pred))

Misclassified samples: 7
Accuracy: 0.844


## Bằng Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
# Đọc file dataset, biến thành các giá trị số
df = pd.read_csv('/content/sms_spam_svm.csv')
y = df.iloc[:, 0].values
y = np.where(y == 'spam', -1, 1)
X = df.iloc[:, [1, 2]].values
# chia dữ liệu thành tập train và tập test
# test_size là 30%
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size=0.3, random_state=0)
# Khởi tạo mô hình hồi quy tuyến tính
lr = LinearRegression()
# Huấn luyện mô hình
lr.fit(X_train, y_train)
# Dự đoán nhãn
y_pred = lr.predict(X_test)
# Làm tròn giá trị dự đoán về -1 hoặc 1
y_pred = [1 if pred >= 0 else -1 for pred in y_pred]
# Đánh giá kết quả
print('Misclassified samples: %d' % (y_test != y_pred).sum())
print('Accuracy: %.3f' % accuracy_score(y_test, y_pred))


Misclassified samples: 7
Accuracy: 0.844


# Bài tập 5: Phát	hiện	Phishing	với	ML - Logistic regession

In [ ]:
import pandas as pd
import numpy as np
from sklearn import *
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
phishing_dataset = np.genfromtxt('/content/drive/MyDrive/Labs/Lab1/phishing_dataset.csv', delimiter=',',dtype=np.int32)
samples = phishing_dataset[:,:-1]
targets = phishing_dataset[:, -1]

from sklearn.model_selection import train_test_split

training_samples, testing_samples, training_targets, testing_targets = train_test_split(samples, targets, test_size=0.2, random_state=0)

log_classifier = LogisticRegression()
log_classifier.fit(training_samples, training_targets)
predictions = log_classifier.predict(testing_samples)

accuracy = 100.0 * accuracy_score(testing_targets, predictions)
print ("Logistic Regression accuracy: " + str(accuracy))

Logistic Regression accuracy: 91.72320217096338


## Chức năng của phương thức genfromtxt() trong thư viện numpy

Tải dữ liệu từ tệp văn bản, xử lý các missing values (là dữ liệu bị khuyết, trống hoặc không tồn tại trong một tập dữ liệu) theo chỉ định.
Một số parameters phổ biến như:
*   **fname:** file, str, pathlib.Path, list of str, generator \
Tệp, tên tệp, danh sách,...Nếu phần mở rộng tên tệp là .gzhoặc .bz2, thì tệp sẽ được giải nén trước.
*   **dtype:** dtype, optional \
Kiểu dữ liệu của mảng kết quả. Nếu Không có, các dtype sẽ được xác định theo nội dung của từng cột, riêng lẻ.
*   **delimiter:** str, int, or sequence, optional \
Chuỗi được sử dụng để phân tách các giá trị. Theo mặc định, bất kỳ khoảng trắng liên tiếp nào cũng hoạt động như dấu phân cách. Một số nguyên hoặc chuỗi số nguyên cũng có thể được cung cấp làm chiều rộng của mỗi trường.



Ví dụ về sử dụng phương thức genfromtxt() trong đoạn mã trên:
*   '/content/drive/MyDrive/Labs/Lab1/phishing_dataset.csv': Đường dẫn đến tệp CSV chứa dữ liệu về phishing

*   delimiter=',': Dữ liệu trong tệp CSV được phân tách bằng dấu phẩy , \
*   dtype=np.int32: Chỉ định kiểu dữ liệu là số nguyên 32-bit (int32)

# Bài tập 6: Sinh	viên hoàn thiện code Decision trees trên và đánh giá kết quả nhận được so với	phương pháp Logistic regression.

In [ ]:
from sklearn import tree
phishing_dataset = np.genfromtxt('/content/phishing_dataset.csv', delimiter=',',dtype=np.int32)
samples = phishing_dataset[:,:-1]
targets = phishing_dataset[:, -1]
training_samples, testing_samples, training_targets, testing_targets = train_test_split(samples, targets, test_size=0.2, random_state=0)

tree_classifier = tree.DecisionTreeClassifier()
tree_classifier.fit(training_samples, training_targets)
predictions = tree_classifier.predict(testing_samples)
accuracy = 100.0 * accuracy_score(testing_targets, predictions)
print ("Decision Tree accuracy: " + str(accuracy))

Decision Tree accuracy: 96.47218453188603


## So sánh với kết quả nhận được của Logistic Regression
- Decision Tree có độ chính xác cao hơn Logistic Regression (96.47% vs. 91.72%). Điều này có thể do Decision Tree có khả năng học tốt các mẫu phức tạp hơn.
- Logistic Regression hoạt động tốt khi dữ liệu có sự phân tách tuyến tính, trong khi Decision Tree linh hoạt hơn với dữ liệu phi tuyến.


# Bài 7: Phát hiện phishing website bằng Logistic Regression và Decision Tree trên dataset từ kaggle

## Logistic Regression

In [ ]:
from sklearn import *
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import pandas as pd

In [ ]:
phishing_1 = pd.read_csv("/content/Phishing_Legitimate_full.csv")
phishing_1

id  NumDots  SubdomainLevel  PathLevel  UrlLength  NumDash  \
0         1        3               1          5         72        0   
1         2        3               1          3        144        0   
2         3        3               1          2         58        0   
3         4        3               1          6         79        1   
4         5        3               0          4         46        0   
...     ...      ...             ...        ...        ...      ...   
9995   9996        3               1          1         50        0   
9996   9997        2               1          4         59        1   
9997   9998        2               1          4         57        0   
9998   9999        3               1          1         49        0   
9999  10000        3               1          2         52        3   

      NumDashInHostname  AtSymbol  TildeSymbol  NumUnderscore  ...  \
0                     0         0            0              0  ...   
1                     0         0            0              2  ...   
2                     0         0            0              0  ...   
3                     0         0            0              0  ...   
4                     0         0            0              0  ...   
...                 ...       ...          ...            ...  ...   
9995                  0         0            0              0  ...   
9996                  0         0            0              0  ...   
9997                  0         0            0              0  ...   
9998                  0         0            0              0  ...   
9999                  0         0            0              0  ...   

      IframeOrFrame  MissingTitle  ImagesOnlyInForm  SubdomainLevelRT  \
0                 0             0                 1                 1   
1                 0             0                 0                 1   
2                 0             0                 0                 1   
3                 0             0                 0                 1   
4                 1             0                 0                 1   
...             ...           ...               ...               ...   
9995              0             0                 0                 1   
9996              1             0                 0                 1   
9997              0             0                 0                 1   
9998              1             0                 0                 1   
9999              0             0                 0                 1   

      UrlLengthRT  PctExtResourceUrlsRT  AbnormalExtFormActionR  \
0               0                     1                       1   
1              -1                     1                       1   
2               0                    -1                       1   
3              -1                     1                       1   
4               1                    -1                       0   
...           ...                   ...                     ...   
9995            1                    -1                       1   
9996            0                     0                       1   
9997            0                     1                       1   
9998            1                     0                       1   
9999            1                     1                       1   

      ExtMetaScriptLinkRT  PctExtNullSelfRedirectHyperlinksRT  CLASS_LABEL  
0                      -1                                   1            1  
1                       1                                   1            1  
2                      -1                                   0            1  
3                       1                                  -1            1  
4                      -1                                  -1            1  
...                   ...                                 ...          ...  
9995                    0                                   1            0  
9996             

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(phishing_1.drop('CLASS_LABEL', axis=1), phishing_1['CLASS_LABEL'], test_size=0.2, random_state=0)

In [ ]:
lr_classifier = LogisticRegression()

In [ ]:
lr_classifier.fit(X_train, y_train)
predictions = lr_classifier.predict(X_test)
accuracy = 100.0 * accuracy_score(y_test, predictions)
print ("Logistic Regression accuracy: " + str(accuracy))

Logistic Regression accuracy: 97.8


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dt_classifier = DecisionTreeClassifier()

In [ ]:
dt_classifier.fit(X_train, y_train)
predictions = dt_classifier.predict(X_test)
accuracy = 100.0 * accuracy_score(y_test, predictions)
print ("Decision Tree accuracy: " + str(accuracy))

Decision Tree accuracy: 100.0


# Bài 8: Phát hiện phishing website bằng LR hoặc Decision Tree trên dataset phishtank

Do phishtank không còn cho phép đăng kí mới tài khoản nên không thể lấy dataset được, nên chúng em đã quyết định thay thế bằng một dataset khác tìm được trên kaggle tại đây: https://www.kaggle.com/datasets/shashwatwork/web-page-phishing-detection-dataset

In [ ]:
phishing_2 = pd.read_csv("/content/dataset_phishing.csv")
phishing_2.head()

url  length_url  \
0              http://www.crestonwood.com/router.php          37   
1  http://shadetreetechnology.com/V4/validation/a...          77   
2  https://support-appleld.com.secureupdate.duila...         126   
3                                 http://rgipt.ac.in          18   
4  http://www.iracing.com/tracks/gateway-motorspo...          55   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_or  ...  \
0               19   0        3           0      0      0       0      0  ...   
1               23   1        1           0      0      0       0      0  ...   
2               50   1        4           1      0      1       2      0  ...   
3               11   0        2           0      0      0       0      0  ...   
4               15   0        2           2      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                0                      1                        0   
1                1                      0                        0   
2                1                      0                        0   
3                1                      0                        0   
4                0                      1                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                          45          -1            0           1   
1                          77        5767            0           0   
2                          14        4004      5828815           0   
3                          62          -1       107721           0   
4                         224        8175         8725           0   

   google_index  page_rank      status  
0             1          4  legitimate  
1             1          2    phishing  
2             1          0    phishing  
3             0          3  legitimate  
4             0          6  legitimate  

[5 rows x 89 columns]

In [ ]:
phishing_2.status = phishing_2.status.map({'legitimate': 0, 'phishing': 1})
phishing_2.head()

url  length_url  \
0              http://www.crestonwood.com/router.php          37   
1  http://shadetreetechnology.com/V4/validation/a...          77   
2  https://support-appleld.com.secureupdate.duila...         126   
3                                 http://rgipt.ac.in          18   
4  http://www.iracing.com/tracks/gateway-motorspo...          55   

   length_hostname  ip  nb_dots  nb_hyphens  nb_at  nb_qm  nb_and  nb_or  ...  \
0               19   0        3           0      0      0       0      0  ...   
1               23   1        1           0      0      0       0      0  ...   
2               50   1        4           1      0      1       2      0  ...   
3               11   0        2           0      0      0       0      0  ...   
4               15   0        2           2      0      0       0      0  ...   

   domain_in_title  domain_with_copyright  whois_registered_domain  \
0                0                      1                        0   
1                1                      0                        0   
2                1                      0                        0   
3                1                      0                        0   
4                0                      1                        0   

   domain_registration_length  domain_age  web_traffic  dns_record  \
0                          45          -1            0           1   
1                          77        5767            0           0   
2                          14        4004      5828815           0   
3                          62          -1       107721           0   
4                         224        8175         8725           0   

   google_index  page_rank  status  
0             1          4       0  
1             1          2       1  
2             1          0       1  
3             0          3       0  
4             0          6       0  

[5 rows x 89 columns]

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(phishing_2.drop(['url', 'status'], axis=1), phishing_2['status'], test_size=0.2, random_state=0)

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
dt_classifier = DecisionTreeClassifier()


In [ ]:
dt_classifier.fit(X_train, y_train)
predictions = dt_classifier.predict(X_test)
accuracy = 100.0 * accuracy_score(y_test, predictions)
print ("Decision Tree accuracy: " + str(accuracy))

Decision Tree accuracy: 94.13823272090988
